ingestion-edge/pubsub_emulator/__init__.py

```diff
@@ -0,0 +1,169 @@
+            self.topics[request.topic].add(subscription)
+        return request
+
+    def DeleteSubscription(self, request, context):  # noqa: D403
+        """DeleteSubscription implementation."""
+        try:
+            subscription = self.subscriptions.pop(request.subscription)
+        except KeyError:
+            context.set_code(grpc.StatusCode.NOT_FOUND)
+        else:
+            for subscriptions in self.topics.values():
+                subscriptions.discard(subscription)
+        return empty_pb2.Empty()
+
+    def Publish(self, request, context):
+        """Publish implementation."""
+        if request.topic in self.status_codes:
+            context.set_code(self.status_codes[request.topic])
+            return pubsub_pb2.PublishResponse()
+        if self.sleep is not None:
+            time.sleep(self.sleep)
+        message_ids = []
+        try:
+            subscriptions = self.topics[request.topic]
+        except KeyError:
+            context.set_code(grpc.StatusCode.NOT_FOUND)
+        else:
+            message_ids = [uuid.uuid4().hex for _ in range(len(request.messages))]
+            for _id, message in zip(message_ids, request.messages):
+                message.message_id = _id
+            for subscription in subscriptions:
+                subscription.published.extend(request.messages)
+        return pubsub_pb2.PublishResponse(message_ids=message_ids)
+
+    def Pull(self, request, context):
+        """Pull implementation."""
+        received_messages = []
+        try:
+            subscription = self.subscriptions[request.subscription]
+        except KeyError:
+            context.set_code(grpc.StatusCode.NOT_FOUND)
+        else:
+            messages = subscription.published[: request.max_messages or 100]
+            subscription.pulled.update(
+                {message.message_id: message for message in messages}
+            )
+            for message in messages:
+                try:
+                    subscription.published.remove(message)
+                except ValueError:
+                    pass
+            received_messages = [
+                pubsub_pb2.ReceivedMessage(ack_id=message.message_id, message=message)
+                for message in messages
+            ]
+        return pubsub_pb2.PullResponse(received_messages=received_messages)
+
+    def Acknowledge(self, request, context):
+        """Acknowledge implementation."""
+        try:
+            subscription = self.subscriptions[request.subscription]
+        except KeyError:
+            context.set_code(grpc.StatusCode.NOT_FOUND)
+        else:
+            for ack_id in request.ack_ids:
+                try:
+                    subscription.pulled.pop(ack_id)
+                except KeyError:
+                    context.set_code(grpc.StatusCode.NOT_FOUND)
+        return empty_pb2.Empty()
+
+    def UpdateTopic(self, request, context):
+        """Repurpose UpdateTopic API for setting up test conditions."""
```

Comment:
Need some more documentation here about what's going on what specific repurposing is going on here, and how to use the setting of statuscode to alter behavior.